components/txn_types/src/lock.rs

```diff
@@ -461,126 +641,180 @@ impl Lock {
-            // Ignore lock when the lock's type is Lock or Pessimistic.
-            return Ok(());
+    pub fn len(&self) -> usize {
+        self.txn_info_segments.len()
+    }
+
+    #[inline]
+    #[must_use]
+    pub fn contains_start_ts(&self, start_ts: TimeStamp) -> bool {
+        self.txn_info_segments.contains_key(&start_ts)
+    }
+
+    pub fn get_lock(&mut self, ts: &TimeStamp) -> Option<&Lock> {
+        if let Some(either) = self.txn_info_segments.get_mut(ts) {
+            match either {
+                Either::Left(encoded) => {
+                    let lock = Lock::parse(encoded).expect("failed to parse shared lock txn info");
+                    *either = Either::Right(lock);
+                    either.as_ref().right()
+                }
+                Either::Right(lock) => Some(lock),
+            }
+        } else {
+            None
         }
+    }
 
-        // The lock is resolved already.
-        if bypass_locks.contains(lock.ts) {
-            return Ok(());
+    // Decode all shared-lock segments so tests can rely on parsed locks.
+    #[cfg(test)]
+    pub fn parse_all(&mut self) {
+        for (_ts, either) in self.txn_info_segments.iter_mut() {
+            if let Either::Left(encoded) = either {
+                let lock = Lock::parse(encoded).expect("failed to parse shared lock txn info");
+                *either = Either::Right(lock);
+            }
         }
+    }
 
-        // Return conflict error.
-        Err(Error::from(ErrorInner::WriteConflict {
-            start_ts: ts,
-            conflict_start_ts: lock.ts,
-            conflict_commit_ts: Default::default(),
-            key: key.to_raw()?,
-            primary: lock.primary.to_vec(),
-            reason: WriteConflictReason::RcCheckTs,
-        }))
+    #[inline]
+    #[must_use]
+    pub fn find_shared_lock_txn(&mut self, start_ts: TimeStamp) -> Option<&Lock> {
+        self.get_lock(&start_ts)
     }
 
-    pub fn check_ts_conflict(
-        lock: Cow<'_, Self>,
-        key: &Key,
-        ts: TimeStamp,
-        bypass_locks: &TsSet,
-        iso_level: IsolationLevel,
-    ) -> Result<()> {
-        match iso_level {
-            IsolationLevel::Si => Lock::check_ts_conflict_si(lock, key, ts, bypass_locks, false),
-            IsolationLevel::RcCheckTs => {
-                Lock::check_ts_conflict_rc_check_ts(lock, key, ts, bypass_locks)
-            }
-            _ => Ok(()),
+    #[inline]
+    #[must_use]
+    pub fn shared_lock_num(&self) -> usize {
+        self.len()
+    }
+
+    fn put_lock(&mut self, ts: TimeStamp, lock: Lock) -> Option<Lock> {
+        match self.txn_info_segments.insert(ts, Either::Right(lock)) {
+            Some(either) => match either {
+                Either::Left(encoded) => {
+                    Some(Lock::parse(&encoded).expect("failed to parse shared lock txn info"))
+                }
+                Either::Right(lock) => Some(lock),
+            },
+            None => None,
```

Comment:
Perhaps more understandable like this, distinguish the existing and new inserting one
```rust
fn put_lock(&mut self, ts: TimeStamp, lock: Lock) -> Option<Lock> {
      let previous = self.txn_info_segments.insert(ts, Either::Right(lock));

      match previous {
          Some(Either::Left(encoded)) => {
              // Previously stored as encoded lock info; decode it back to a Lock.
              Some(Lock::parse(&encoded).expect("failed to parse shared lock txn info"))
          }
          Some(Either::Right(existing_lock)) => Some(existing_lock),
          None => None,
      }
  }
```